File: exp-09/q1-matrix-chain-multiplication/solution.js
const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });

const MAX = 100;
const s = Array.from({ length: MAX }, () => new Array(MAX).fill(0));
let tokens = [];

function printOrder(i, j) {
    if (i === j) {
        process.stdout.write(`A${i}`);
        return;
    }
    process.stdout.write('(');
    printOrder(i, s[i][j]);
    printOrder(s[i][j] + 1, j);
    process.stdout.write(')');
}

rl.on('line', (line) => {
    tokens = tokens.concat(line.trim().split(/\s+/).map(Number));
});

rl.on('close', () => {
    const n = tokens[0];
    const p = tokens.slice(1, n + 2);
    const m = Array.from({ length: MAX }, () => new Array(MAX).fill(0));

    for (let length = 2; length <= n; length++) {
        for (let i = 1; i <= n - length + 1; i++) {
            const j = i + length - 1;
            m[i][j] = Infinity;
            for (let k = i; k < j; k++) {
                const cost = m[i][k] + m[k + 1][j] + p[i - 1] * p[k] * p[j];
                if (cost < m[i][j]) {
                    m[i][j] = cost;
                    s[i][j] = k;
                }
            }
        }
    }

    console.log(`Minimum cost: ${m[1][n]}`);
    process.stdout.write('Optimal parenthesization: ');
    printOrder(1, n);
    console.log();
});

File: exp-09/q1-matrix-chain-multiplication/solution.py
import sys

MAX = 100
s = [[0] * MAX for _ in range(MAX)]


def print_order(i, j):
    if i == j:
        print(f"A{i}", end="")
        return
    print("(", end="")
    print_order(i, s[i][j])
    print_order(s[i][j] + 1, j)
    print(")", end="")


def main():
    data = list(map(int, sys.stdin.read().split()))
    n = data[0]
    p = data[1:n + 2]
    m = [[0] * MAX for _ in range(MAX)]

    for length in range(2, n + 1):
        for i in range(1, n - length + 2):
            j = i + length - 1
            m[i][j] = float("inf")
            for k in range(i, j):
                cost = m[i][k] + m[k + 1][j] + p[i - 1] * p[k] * p[j]
                if cost < m[i][j]:
                    m[i][j] = cost
                    s[i][j] = k

    print(f"Minimum cost: {m[1][n]}")
    print("Optimal parenthesization: ", end="")
    print_order(1, n)
    print()


if __name__ == "__main__":
    main()
